src/EventListener/TypesenseIndexer.php

```diff
@@ -2,88 +2,161 @@
+    private $documentsToIndex = [];
+    private $documentsToUpdate = [];
+    private $documentsToDelete = [];
+
+    public function __construct(
+        CollectionManager $collectionManager,
+        DocumentManager $documentManager,
+        DoctrineToTypesenseTransformer $transformer
+    ) {
         $this->collectionManager = $collectionManager;
-        $this->documentManager = $documentManager;
-        $this->transformer = $transformer;
+        $this->documentManager   = $documentManager;
+        $this->transformer       = $transformer;
 
-        $this->managedClassNames = $this->collectionManager->getManagedClassNames();
+        $this->managedClassNames  = $this->collectionManager->getManagedClassNames();
+        $this->objectsIDsToDelete = [];
     }
 
     public function postPersist(LifecycleEventArgs $args)
     {
-        $entity = $args->getObject();
-        $entityClassname = ClassUtils::getClass($entity);
+        $entity          = $args->getObject();
 
-        if (!in_array($entityClassname, array_values($this->managedClassNames))) {
+        if ($this->entityIsNotManaged($entity)) {
             return;
         }
 
-        $collection = array_search($entityClassname, $this->managedClassNames);
-        $data = $this->transformer->convert($entity);
-        $this->documentManager->index($collection, $data);
-    }
+        $collection = $this->getCollectionName($entity);
+        $data       = $this->transformer->convert($entity);
 
+        $this->documentsToIndex[] = [$collection, $data];
+    }
 
     public function postUpdate(LifecycleEventArgs $args)
     {
-        $entity = $args->getObject();
-        $entityClassname = ClassUtils::getClass($entity);
-        if (!in_array($entityClassname, array_values($this->managedClassNames))) {
+        $entity          = $args->getObject();
+
+        if ($this->entityIsNotManaged($entity)) {
             return;
         }
 
-        $collectionDefinitionKey = array_search($entityClassname, $this->managedClassNames);
-        $collectionConfig = $this->collectionManager->getCollectionDefinitions()[$collectionDefinitionKey];
+        $collectionDefinitionKey = $this->getCollectionName($entity);
+        $collectionConfig        = $this->collectionManager->getCollectionDefinitions()[$collectionDefinitionKey];
 
+        $this->checkPrimaryKeyExists($collectionConfig);
 
-        $primaryKey = $this->getPrimaryKeyInfo($collectionConfig);
+        $collection = $this->getCollectionName($entity);
+        $data       = $this->transformer->convert($entity);
 
-        $collection = array_search($entityClassname, $this->managedClassNames);
-        $data = $this->transformer->convert($entity);
-   
-        $this->documentManager->delete($collection, $data['id']);
-        $this->documentManager->index($collection, $data);
+        $this->documentsToUpdate[] = [$collection, $data['id'], $data];
     }
 
+    private function checkPrimaryKeyExists($collectionConfig)
```

Comment:
Hello, by design, the primary key exists automatically (it is added if not declared in src/DependencyInjection/ACSEOTypesenseExtension.php:111).

is there a reason why you added this check ?

Reply:
You're right !

I added it because of the old `getPrimaryKeyInfo` function : https://github.com/acseo/TypesenseBundle/blob/cb6c55da28959592a62bd99ea4b7071fcae6a210/src/EventListener/TypesenseIndexer.php#L80 but I saw that the variable `$primaryKey`is not used so I only did the check.
But you're right : we can remove this check.